crates/stark_hash/src/felt.rs

```diff
@@ -0,0 +1,762 @@
+            // substraction overflow; input is smaller than modulus
+            Ok(Felt(bytes))
+        }
+    }
+
+    /// Returns a bit view of the 251 least significant bits in MSB order.
+    pub fn view_bits(&self) -> &BitSlice<Msb0, u8> {
+        &self.0.view_bits()[5..]
+    }
+
+    /// Creates a [Felt] from up-to 251 bits.
+    pub fn from_bits(bits: &BitSlice<Msb0, u8>) -> Result<Self, OverflowError> {
+        if bits.len() > 251 {
+            return Err(OverflowError);
+        }
+
+        let mut bytes = [0u8; 32];
+        bytes.view_bits_mut::<Msb0>()[256 - bits.len()..].copy_from_bitslice(bits);
+
+        Ok(Self(bytes))
+    }
+
+    /// Returns `true` if the value of [`Felt`] is larger than `2^251 - 1`.
+    ///
+    /// Every [`Felt`] that is used to traverse a Merkle-Patricia Tree
+    /// must not exceed 251 bits, since 251 is the height of the tree.
+    pub const fn has_more_than_251_bits(&self) -> bool {
+        self.0[0] & 0b1111_1000 > 0
+    }
+
+    pub const fn from_u64(u: u64) -> Self {
+        const_expect!(
+            Self::from_be_slice(&u.to_be_bytes()),
+            "64 bits is less than 251 bits"
+        )
+    }
+
+    pub const fn from_u128(u: u128) -> Self {
+        const_expect!(
+            Self::from_be_slice(&u.to_be_bytes()),
+            "128 bits is less than 251 bits"
+        )
+    }
+}
+
+macro_rules! const_expect {
+    ($e:expr, $why:expr) => {{
+        match $e {
+            Ok(x) => x,
+            Err(_) => panic!(concat!("Expectation failed: ", $why)),
+        }
+    }};
+}
+
+use const_expect;
+
+impl From<u64> for Felt {
+    fn from(value: u64) -> Self {
+        Self::from_u64(value)
+    }
+}
+
+impl From<u128> for Felt {
+    fn from(value: u128) -> Self {
+        Self::from_u128(value)
+    }
+}
+
+impl std::ops::Add for Felt {
+    type Output = Felt;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        let result = FieldElement::from(self) + FieldElement::from(rhs);
+        Felt::from(result)
+    }
+}
+
+impl From<Felt> for FieldElement {
+    fn from(hash: Felt) -> Self {
+        debug_assert_eq!(
+            std::mem::size_of::<FieldElement>(),
+            std::mem::size_of::<Felt>()
+        );
+        Self::from_repr(FieldElementRepr(hash.to_be_bytes())).unwrap()
+    }
+}
+
+impl From<FieldElement> for Felt {
+    fn from(fp: FieldElement) -> Self {
+        debug_assert_eq!(
+            std::mem::size_of::<FieldElement>(),
+            std::mem::size_of::<Felt>()
+        );
+        // unwrap is safe because the FieldElement and StarkHash
+        // should both be smaller than the field modulus.
+        Felt::from_be_bytes(fp.to_repr().0).unwrap()
+    }
+}
+
+impl Felt {
+    /// A convenience function which parses a hex string into a [Felt].
+    ///
+    /// Supports both upper and lower case hex strings, as well as an
+    /// optional "0x" prefix.
+    pub const fn from_hex_str(hex_str: &str) -> Result<Self, HexParseError> {
```

Comment:
I guess [this one](https://github.com/eqlabs/pathfinder/blob/db5987f388d53897574e50c0a9c5900761bb2bd9/crates/serde/src/lib.rs#L329) could be constified too then?

Reply:
Probably yes; I only tackled the macro related one because it was using something else.